src/main/kotlin/kotlinx/kover/KoverPlugin.kt

```diff
@@ -19,74 +32,192 @@ class KoverPlugin : Plugin<Project> {
+        val intellijAgent = target.createIntellijAgent(koverExtension)
+        val jacocoAgent = target.createJacocoAgent(koverExtension)
 
-        val intellijConfig = target.createIntellijConfig(koverExtension)
-        val jacocoConfig = target.createJacocoConfig(koverExtension)
-        val jacocoAgent = JacocoAgent(jacocoConfig, target)
-        val intellijAgent = IntellijAgent(intellijConfig)
-
-        target.tasks.withType(Test::class.java).configureEach {
-            it.applyCoverage(intellijAgent, jacocoAgent)
+        target.allprojects {
+            it.applyToProject(koverExtension, intellijAgent, jacocoAgent)
         }
+        target.createCollectingTask()
     }
 
-    private fun Test.applyCoverage(intellijAgent: IntellijAgent, jacocoAgent: JacocoAgent) {
-        val taskExtension = extensions.create("kover", KoverTaskExtension::class.java, project.objects)
+    private fun Project.createCollectingTask() {
+        tasks.register(COLLECT_TASK_NAME, Copy::class.java) {
+            it.group = TASK_GROUP
+            subprojects { sub ->
+                val xmlReportTask = sub.tasks.withType(KoverXmlReportTask::class.java).getByName(XML_REPORT_TASK_NAME)
+                val htmlReportTask =
+                    sub.tasks.withType(KoverHtmlReportTask::class.java).getByName(HTML_REPORT_TASK_NAME)
 
-        taskExtension.xmlReportFile.set(this.project.provider {
-            project.layout.buildDirectory.get().file("reports/kover/$name.xml").asFile
-        })
+                it.mustRunAfter(xmlReportTask)
+                it.mustRunAfter(htmlReportTask)
 
-        taskExtension.htmlReportDir.set(this.project.provider {
-            project.layout.buildDirectory.get().dir("reports/kover/html/$name")
-        })
+                it.from(xmlReportTask.xmlReportFile)
 
-        taskExtension.binaryReportFile.set(this.project.provider {
-            val suffix = if (taskExtension.coverageEngine == CoverageEngine.JACOCO) ".exec" else ".ic"
-            project.layout.buildDirectory.get().file("kover/$name$suffix").asFile
-        })
+                it.from(htmlReportTask.htmlReportDir) { sp ->
+                    sp.into("html/${sub.name}/")
+                }
 
-        jvmArgumentProviders.add(CoverageArgumentProvider(jacocoAgent, intellijAgent, taskExtension, this))
+                it.into(layout.buildDirectory.dir("reports/kover"))
+            }
+        }
+    }
 
-        doLast {
-            taskExtension.generateXml = taskExtension.generateXml
-                    // turn on XML report for intellij agent if verification rules are defined
-                    || (taskExtension.coverageEngine == CoverageEngine.INTELLIJ && taskExtension.rules.isNotEmpty())
 
-            if (!(taskExtension.isEnabled && (taskExtension.generateXml || taskExtension.generateHtml || taskExtension.rules.isNotEmpty()))) {
-                return@doLast
-            }
+    private fun Project.applyToProject(
```

Comment:
All tasks have to have a clear description.
Now it seems kinda off-rails

![image](https://user-images.githubusercontent.com/2780116/139707600-4fe01aa0-e72a-43d3-9d71-c81db2c442bf.png)